File: lib/services/alert_engine.dart
import 'dart:async';
import 'package:vibration/vibration.dart';

import '../models/types.dart';
import 'audio_engine.dart';

typedef LevelCallback = void Function(AlertLevel level);
typedef EventCallback = void Function(TripEvent event);
typedef MsCallback = void Function(int ms);
typedef DigitCallback = void Function(String digit, int index);
typedef IntCallback = void Function(int value);

class AlertEngine {
  static const _sustainMs = 500;
  static const _cooldownMs = 3000;
  static const _drowsyWindowMs = 30000;
  static const _drowsyThreshold = 3;
  static const _calibrationMs = 3000;
  static const _faceLostToleranceMs = 3000;

  static const _blinkIgnoreMs = 800;
  static const _warningAtMs = 5000;
  static const _criticalAtMs = 10000;
  static const _emergencyAtMs = 15000;
  static const _eyeOpenResetMs = 1000;

  final AudioEngine audio;
  double confidenceThreshold;
  String emergencyNumber;

  final LevelCallback onLevel;
  final EventCallback onEvent;
  final MsCallback onClosedMs;
  final DigitCallback onDialerDigit;
  final IntCallback onCountdown;
  final void Function() onCallingStarted;

  AlertLevel _level = AlertLevel.none;
  AlertLevel get level => _level;

  int _graceUntilMs = 0;

  // Track A
  int _yawnSustainStart = 0;
  int _headDownSustainStart = 0;
  int _lastYawnEventAt = 0;
  int _lastHeadDownEventAt = 0;
  final List<TripEvent> _events = [];

  // Track B
  int _closedSince = 0;
  int _openSince = 0;
  int _faceLostSince = 0;
  bool _inEmergencyFlow = false;

  Timer? _countdownTimer;

  AlertEngine({
    required this.audio,
    required this.confidenceThreshold,
    required this.emergencyNumber,
    required this.onLevel,
    required this.onEvent,
    required this.onClosedMs,
    required this.onDialerDigit,
    required this.onCountdown,
    required this.onCallingStarted,
  });

  void start() {
    final now = DateTime.now().millisecondsSinceEpoch;
    _graceUntilMs = now + _calibrationMs;
    _events.clear();
    _level = AlertLevel.none;
    _closedSince = 0;
    _openSince = 0;
    _inEmergencyFlow = false;
  }

  Future<void> stop() async {
    _countdownTimer?.cancel();
    _countdownTimer = null;
    await audio.stopAll();
    _setLevel(AlertLevel.none);
    _inEmergencyFlow = false;
  }

  Future<void> dismiss() async {
    _countdownTimer?.cancel();
    _countdownTimer = null;
    await audio.stopAll();
    _events.clear();
    _closedSince = 0;
    _openSince = 0;
    _inEmergencyFlow = false;
    _graceUntilMs = DateTime.now().millisecondsSinceEpoch + 10000;
    _setLevel(AlertLevel.none);
  }

  Future<void> ingest(DetectionResult result) async {
    final now = result.tsMs;
    final inGrace = now < _graceUntilMs;

    if (result.faceLost) {
      if (_faceLostSince == 0) _faceLostSince = now;
      if (now - _faceLostSince > _faceLostToleranceMs) onClosedMs(0);
      return;
    }
    _faceLostSince = 0;

    if (inGrace) {
      onClosedMs(0);
      return;
    }

    final face = result.faces.first;

    // ---- Track A — yawn and head-down are INDEPENDENT signals.
    // Use the per-signal binary confidence, not the combined `face.conf`,
    // so a yawning forward-facing user still registers as yawn.
    final yawnPasses = face.isYawn && face.yawnConf >= confidenceThreshold;
    final headDownPasses =
        face.isHeadDown && face.headPoseConf >= confidenceThreshold;

    if (yawnPasses) {
      if (_yawnSustainStart == 0) _yawnSustainStart = now;
      if (now - _yawnSustainStart >= _sustainMs &&
          now - _lastYawnEventAt >= _cooldownMs) {
        _lastYawnEventAt = now;
        await _registerEvent(TripEvent(now, EventType.yawn));
      }
    } else {
      _yawnSustainStart = 0;
    }

    if (headDownPasses) {
      if (_headDownSustainStart == 0) _headDownSustainStart = now;
      if (now - _headDownSustainStart >= _sustainMs &&
          now - _lastHeadDownEventAt >= _cooldownMs) {
        _lastHeadDownEventAt = now;
        await _registerEvent(TripEvent(now, EventType.headDown));
      }
    } else {
      _headDownSustainStart = 0;
    }

    _events.removeWhere((e) =>
        now - e.ts > _drowsyWindowMs ||
        (e.type != EventType.yawn && e.type != EventType.headDown));

    // ---- Track B ----
    final closed = _classifyEyesClosed(face);
    if (closed) {
      if (_closedSince == 0) _closedSince = now;
      _openSince = 0;
    } else {
      if (_openSince == 0) _openSince = now;
      if (now - _openSince >= _eyeOpenResetMs) {
        if (_closedSince != 0) {
          // Lock past 10s: once we've escalated to CRITICAL or EMERGENCY, only
          // a manual Cancel/dismiss can stop the alarm. Eye-open won't reset.
          final locked = _level == AlertLevel.critical ||
              _level == AlertLevel.emergency;
          if (!locked) {
            _countdownTimer?.cancel();
            _countdownTimer = null;
            await audio.stopAll();
            _inEmergencyFlow = false;
            _closedSince = 0;
          }
        } else {
          _closedSince = 0;
        }
      }
    }

    final closedMs = _closedSince == 0 ? 0 : now - _closedSince;
    onClosedMs(closedMs);
    await _updateLevel(closedMs);
  }

  bool _classifyEyesClosed(FacePrediction face) {
    final eyes = face.eyes.where((e) => e.conf >= confidenceThreshold).toList();
    if (eyes.isEmpty) return true; // face but no confident eyes → likely closed
    return eyes.every((e) => e.eyeClass == EyeClass.closed);
  }

  Future<void> _registerEvent(TripEvent ev) async {
    _events.add(ev);
    onEvent(ev);
    await audio.playBuzz();
    if (await Vibration.hasVibrator() ?? false) {
      Vibration.vibrate(duration: 300);
    }

    if (_events.length >= _drowsyThreshold &&
        _level != AlertLevel.critical &&
        _level != AlertLevel.emergency) {
      onEvent(TripEvent(ev.ts, EventType.drowsy));
      _setLevel(AlertLevel.drowsy);
      await audio.startPullover();
    }
  }

  Future<void> _updateLevel(int closedMs) async {
    if (closedMs >= _emergencyAtMs) {
      if (_level != AlertLevel.emergency) {
        _setLevel(AlertLevel.emergency);
        onEvent(TripEvent(
            DateTime.now().millisecondsSinceEpoch, EventType.emergency));
        await _startEmergencyFlow();
      }
      return;
    }
    if (closedMs >= _criticalAtMs) {
      if (_level != AlertLevel.critical) {
        _setLevel(AlertLevel.critical);
        onEvent(TripEvent(
            DateTime.now().millisecondsSinceEpoch, EventType.critical));
        await audio.startSiren();
        _startCountdown();
      }
      return;
    }
    if (closedMs >= _warningAtMs) {
      if (_level != AlertLevel.warning) {
        _setLevel(AlertLevel.warning);
        await audio.startPullover();
      }
      return;
    }
    if (closedMs >= _blinkIgnoreMs) {
      if (_level == AlertLevel.none) _setLevel(AlertLevel.eyesClosing);
      return;
    }
    if (closedMs == 0 &&
        (_level == AlertLevel.warning ||
            _level == AlertLevel.critical ||
            _level == AlertLevel.emergency ||
            _level == AlertLevel.eyesClosing)) {
      _countdownTimer?.cancel();
      _countdownTimer = null;
      await audio.stopAll();
      _inEmergencyFlow = false;
      if (_events.length >= _drowsyThreshold) {
        _setLevel(AlertLevel.drowsy);
        await audio.startPullover();
      } else {
        _setLevel(AlertLevel.none);
      }
    }
  }

  void _setLevel(AlertLevel l) {
    if (_level == l) return;
    _level = l;
    onLevel(l);
  }

  void _startCountdown() {
    var n = 5;
    onCountdown(n);
    _countdownTimer?.cancel();
    _countdownTimer = Timer.periodic(const Duration(seconds: 1), (t) {
      n -= 1;
      onCountdown(n < 0 ? 0 : n);
      if (n <= 0) {
        t.cancel();
        _countdownTimer = null;
      }
    });
  }

  Future<void> _startEmergencyFlow() async {
    if (_inEmergencyFlow) return;
    _inEmergencyFlow = true;
    await audio.stopPullover();
    await audio.duckSiren();

    final offsets = await audio.playDialer();
    final number = emergencyNumber;
    for (var i = 0; i < offsets.length && i < number.length; i++) {
      Future.delayed(Duration(milliseconds: offsets[i]), () {
        onDialerDigit(number[i], i);
      });
    }

    audio.onDialerEnd(() async {
      onCallingStarted();
      await audio.playCallingTimes(3);
    });
  }
}

File: webapp/src/lib/alertEngine.ts
// Browser port of lib/services/alert_engine.dart.
// Two independent state machines (focus reminders + critical eyes-closed)
// driven off the per-frame DetectionResult stream. Behavior matches DESIGN.md §4.

import { AudioEngine } from "./audioEngine";
import { dlog } from "./debug";
import {
  AlertLevel,
  DetectionResult,
  FacePrediction,
  TripEvent,
} from "./types";

const SUSTAIN_MS = 500;
const COOLDOWN_MS = 3000;
const DROWSY_WINDOW_MS = 30000;
const DROWSY_THRESHOLD = 3;
const CALIBRATION_MS = 3000;
const FACE_LOST_TOLERANCE_MS = 3000;

const BLINK_IGNORE_MS = 800;
const WARNING_AT_MS = 5000;
const CRITICAL_AT_MS = 10000;
const EMERGENCY_AT_MS = 15000;
const EYE_OPEN_RESET_MS = 1000;

export interface AlertCallbacks {
  onLevel: (level: AlertLevel) => void;
  onEvent: (event: TripEvent) => void;
  onClosedMs: (ms: number) => void;
  onDrowsyCount: (count: number) => void;
  onDialerDigit: (digit: string, index: number) => void;
  onCountdown: (value: number) => void;
  onCallingStarted: () => void;
}

function vibrate(ms: number) {
  if (typeof navigator !== "undefined" && navigator.vibrate) {
    try {
      navigator.vibrate(ms);
    } catch {
      /* not supported */
    }
  }
}

export class AlertEngine {
  confidenceThreshold: number;
  emergencyNumber: string;
  private audio: AudioEngine;
  private cb: AlertCallbacks;

  private _level = AlertLevel.none;
  get level() {
    return this._level;
  }

  private graceUntilMs = 0;

  // Track A
  private yawnSustainStart = 0;
  private headDownSustainStart = 0;
  private lastYawnEventAt = 0;
  private lastHeadDownEventAt = 0;
  private events: TripEvent[] = [];
  private lastCount = -1;

  // Track B
  private closedSince = 0;
  private openSince = 0;
  private faceLostSince = 0;
  private inEmergencyFlow = false;

  private countdownTimer: ReturnType<typeof setInterval> | null = null;

  constructor(
    audio: AudioEngine,
    confidenceThreshold: number,
    emergencyNumber: string,
    cb: AlertCallbacks,
  ) {
    this.audio = audio;
    this.confidenceThreshold = confidenceThreshold;
    this.emergencyNumber = emergencyNumber;
    this.cb = cb;
  }

  start() {
    const now = Date.now();
    this.graceUntilMs = now + CALIBRATION_MS;
    this.events = [];
    this._level = AlertLevel.none;
    this.closedSince = 0;
    this.openSince = 0;
    this.inEmergencyFlow = false;
    this.emitCount();
  }

  stop() {
    this.clearCountdown();
    this.audio.stopAll();
    this.setLevel(AlertLevel.none);
    this.inEmergencyFlow = false;
    this.events = [];
    this.emitCount();
  }

  dismiss() {
    this.clearCountdown();
    this.audio.stopAll();
    this.events = [];
    this.closedSince = 0;
    this.openSince = 0;
    this.inEmergencyFlow = false;
    this.graceUntilMs = Date.now() + 10000;
    this.setLevel(AlertLevel.none);
    this.emitCount();
  }

  async ingest(result: DetectionResult) {
    const now = result.tsMs;
    const inGrace = now < this.graceUntilMs;

    if (result.faceLost) {
      if (this.faceLostSince === 0) this.faceLostSince = now;
      if (now - this.faceLostSince > FACE_LOST_TOLERANCE_MS) this.cb.onClosedMs(0);
      return;
    }
    this.faceLostSince = 0;

    if (inGrace) {
      this.cb.onClosedMs(0);
      return;
    }

    const face = result.faces[0];

    // ---- Track A — yawn / head-down are INDEPENDENT signals ----
    const yawnPasses = face.isYawn && face.yawnConf >= this.confidenceThreshold;
    const headDownPasses =
      face.isHeadDown && face.headPoseConf >= this.confidenceThreshold;

    if (yawnPasses) {
      if (this.yawnSustainStart === 0) this.yawnSustainStart = now;
      if (
        now - this.yawnSustainStart >= SUSTAIN_MS &&
        now - this.lastYawnEventAt >= COOLDOWN_MS
      ) {
        this.lastYawnEventAt = now;
        await this.registerEvent({ ts: now, type: "yawn" });
      }
    } else {
      this.yawnSustainStart = 0;
    }

    if (headDownPasses) {
      if (this.headDownSustainStart === 0) this.headDownSustainStart = now;
      if (
        now - this.headDownSustainStart >= SUSTAIN_MS &&
        now - this.lastHeadDownEventAt >= COOLDOWN_MS
      ) {
        this.lastHeadDownEventAt = now;
        await this.registerEvent({ ts: now, type: "headDown" });
      }
    } else {
      this.headDownSustainStart = 0;
    }

    this.events = this.events.filter(
      (e) =>
        now - e.ts <= DROWSY_WINDOW_MS &&
        (e.type === "yawn" || e.type === "headDown"),
    );
    this.emitCount();

    // ---- Track B ----
    const closed = this.classifyEyesClosed(face);
    if (closed) {
      if (this.closedSince === 0) this.closedSince = now;
      this.openSince = 0;
    } else {
      if (this.openSince === 0) this.openSince = now;
      if (now - this.openSince >= EYE_OPEN_RESET_MS) {
        if (this.closedSince !== 0) {
          const locked =
            this._level === AlertLevel.critical ||
            this._level === AlertLevel.emergency;
          if (!locked) {
            this.clearCountdown();
            this.audio.stopAll();
            this.inEmergencyFlow = false;
            this.closedSince = 0;
          }
        } else {
          this.closedSince = 0;
        }
      }
    }

    const closedMs = this.closedSince === 0 ? 0 : now - this.closedSince;
    this.cb.onClosedMs(closedMs);
    await this.updateLevel(closedMs);
  }

  private classifyEyesClosed(face: FacePrediction): boolean {
    const eyes = face.eyes.filter((e) => e.conf >= this.confidenceThreshold);
    if (eyes.length === 0) return true; // face but no confident eyes → likely closed
    return eyes.every((e) => e.eyeClass === "Closed");
  }

  // Surface the live count of yawn / head-down events inside the rolling 30s
  // window (capped at the drowsy threshold) so the UI can show "x/3". Deduped so
  // the callback only fires when the displayed value actually changes.
  private emitCount() {
    const count = Math.min(this.events.length, DROWSY_THRESHOLD);
    if (count !== this.lastCount) {
      this.lastCount = count;
      this.cb.onDrowsyCount(count);
    }
  }

  private async registerEvent(ev: TripEvent) {
    this.events.push(ev);
    this.cb.onEvent(ev);
    await this.audio.playBuzz();
    vibrate(300);

    if (
      this.events.length >= DROWSY_THRESHOLD &&
      this._level !== AlertLevel.critical &&
      this._level !== AlertLevel.emergency
    ) {
      this.cb.onEvent({ ts: ev.ts, type: "drowsy" });
      this.setLevel(AlertLevel.drowsy);
      await this.audio.startPullover();
    }
  }

  private async updateLevel(closedMs: number) {
    if (closedMs >= EMERGENCY_AT_MS) {
      if (this._level !== AlertLevel.emergency) {
        this.setLevel(AlertLevel.emergency);
        this.cb.onEvent({ ts: Date.now(), type: "emergency" });
        await this.startEmergencyFlow();
      }
      return;
    }
    if (closedMs >= CRITICAL_AT_MS) {
      if (this._level !== AlertLevel.critical) {
        this.setLevel(AlertLevel.critical);
        this.cb.onEvent({ ts: Date.now(), type: "critical" });
        await this.audio.startSiren();
        this.startCountdown();
      }
      return;
    }
    if (closedMs >= WARNING_AT_MS) {
      if (this._level !== AlertLevel.warning) {
        this.setLevel(AlertLevel.warning);
        await this.audio.startPullover();
      }
      return;
    }
    if (closedMs >= BLINK_IGNORE_MS) {
      if (this._level === AlertLevel.none) this.setLevel(AlertLevel.eyesClosing);
      return;
    }
    if (
      closedMs === 0 &&
      (this._level === AlertLevel.warning ||
        this._level === AlertLevel.critical ||
        this._level === AlertLevel.emergency ||
        this._level === AlertLevel.eyesClosing)
    ) {
      this.clearCountdown();
      this.audio.stopAll();
      this.inEmergencyFlow = false;
      if (this.events.length >= DROWSY_THRESHOLD) {
        this.setLevel(AlertLevel.drowsy);
        await this.audio.startPullover();
      } else {
        this.setLevel(AlertLevel.none);
      }
    }
  }

  private setLevel(l: AlertLevel) {
    if (this._level === l) return;
    this._level = l;
    this.cb.onLevel(l);
  }

  private clearCountdown() {
    if (this.countdownTimer) {
      clearInterval(this.countdownTimer);
      this.countdownTimer = null;
    }
  }

  private startCountdown() {
    let n = 5;
    this.cb.onCountdown(n);
    this.clearCountdown();
    this.countdownTimer = setInterval(() => {
      n -= 1;
      this.cb.onCountdown(n < 0 ? 0 : n);
      if (n <= 0) this.clearCountdown();
    }, 1000);
  }

  private async startEmergencyFlow() {
    dlog(`EMERGENCY flow start (already=${this.inEmergencyFlow})`);
    if (this.inEmergencyFlow) return;
    this.inEmergencyFlow = true;
    this.audio.stopPullover();
    this.audio.duckSiren();

    const offsets = await this.audio.playDialer();
    const number = this.emergencyNumber;
    for (let i = 0; i < offsets.length && i < number.length; i++) {
      setTimeout(() => this.cb.onDialerDigit(number[i], i), offsets[i]);
    }

    this.audio.onDialerEnd(async () => {
      this.cb.onCallingStarted();
      await this.audio.playCallingTimes(3);
    });
  }
}
